python/src/iceberg/types.py

```diff
@@ -82,7 +122,23 @@ def scale(self) -> int:
         return self._scale
 
 
-class NestedField(object):
+class NestedField(Type):
+    """equivalent of `NestedField` type from Java implementation"""
+
+    _implemented: Dict[Tuple[bool, int, str, Type, Optional[str]], "NestedField"] = {}
+
+    def __new__(
+        cls,
+        is_optional: bool,
+        field_id: int,
+        name: str,
+        field_type: Type,
+        doc: Optional[str] = None,
+    ):
+        key = is_optional, field_id, name, field_type, doc
+        cls._implemented[key] = cls._implemented.get(key, object.__new__(cls))
```

Comment:
In this line, will it always create a new object, `object.__new__(cls)` and pass it to get method call?

Reply:
Yeah you're right. Unnecessarily inefficient. changed it now to use `or` instead